Add AddOpenTelemetrySelfDiagnosticsLogging to OpenTelemetry.Extensions.Hosting

ejsmith opened the pull request:
Makes it easy to output self-diagnostics events to the .NET logging system

ejsmith commented on src/OpenTelemetry.Extensions.Hosting/Implementation/SelfDiagnosticsEventLogForwarder.cs:
```diff
@@ -0,0 +1,160 @@
+        /// <summary>
+        /// This method records the events from event sources to a local file, which is provided as a stream object by
+        /// SelfDiagnosticsConfigRefresher class. The file size is bound to a upper limit. Once the write position
+        /// reaches the end, it will be reset to the beginning of the file.
```
Updated

CodeBlanch commented:
Funny I did the same thing. Here's mine: https://github.com/Macross-Software/core/tree/develop/ClassLibraries/Macross.OpenTelemetry.Extensions

ejsmith commented:
Nice! I think it's really important to be able to easily enable these to go to the normal logging system and see why something isn't working or even just how it's working.

reyang commented on src/OpenTelemetry.Extensions.Hosting/Implementation/EventSourceEventFormatting.cs:
```diff
@@ -0,0 +1,89 @@
+                }
+            }
+
+            var stringBuilder = new StringBuilder();
```
Consider caching the StringBuilder + providing hint of the allocation?
https://www.dotnetperls.com/stringbuilder-cache

ejsmith replied:
Updated to use cached stringbuilder

reyang commented on src/OpenTelemetry.Extensions.Hosting/Implementation/SelfDiagnosticsEventLogForwarder.cs:
```diff
@@ -0,0 +1,158 @@
+                case EventLevel.LogAlways:
+                    return LogLevel.Information;
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(level), level, null);
```
Throw or swallow?

ejsmith replied:
Removed throw and just set to LogLevel.None.

reyang commented on src/OpenTelemetry.Extensions.Hosting/Implementation/SelfDiagnosticsEventLogForwarder.cs:
```diff
@@ -0,0 +1,158 @@
+    /// SelfDiagnosticsEventListener class enables the events from OpenTelemetry event sources
+    /// and write the events to an ILoggerFactory.
+    /// </summary>
+    internal class SelfDiagnosticsEventLogForwarder : EventListener
```
Is this intended for local debugging only, or for production use?
I'll be a bit concerned if it is for production:
1. The pattern matching "OpenTelemetry-*" and putting things in the `loggers` dictionary could be an attack surface, causing unbounded memory usage.
2. If folks consumes the logs in OpenTelemetry processor/exporters, it might dead loop (e.g. self-diagnostics go to logs, logs go to processor, processor triggered self-diagnostics).

ejsmith replied:
I would think it would just be for troubleshooting setups. Don't really think you'd ever want to have it on in production. Maybe briefly to figure out a telementry issue.

reyang replied:
#1258 has captured some of the requirements for a production-ready self-diagnostics solution.

If this feature is intended for non-production troubleshooting, is there a way we can clearly indicate that so people won't misuse it (e.g. a separate package, similar like the ConsoleExporter)? (e.g. in C++ folks normally do debug/release bits, doesn't seem to be a common practice in C# although technically it is supported)

ejsmith replied:
I mean, it's just listening to the event sources and writing to the logging system. Both of those things happen in production. I guess you might leave it on with a high minimum level like you do normal logging. So you can see errors and such.

ejsmith replied:
Also, @reyang we wouldn't need any of this if the whole project would just make use of MS.Ext.Logging like I was trying to argue for in that issue. That hasn't happened and the system currently uses event sources. So this is just forwarding those events to the standard logging system that every modern .NET app uses.

reyang replied:
And you know that we cannot use `Microsoft.Extensions.Logging` due to https://github.com/open-telemetry/opentelemetry-dotnet#supported-net-versions.

cijothomas replied:
yes. I remember Eric asking for ILogger in the originals PRs for self-diag.

Could we explore adding multiple destinations for SelfDiagnostic feature, File being one of them, ILogger being another (conditionally support where ILogger is supported).

reyang replied:
My biggest concern here is that we could easily end up with dead loop.

ejsmith replied:
@cijothomas that would require moving away from event sources being the primary thing that OTel is writing to internally.

cijothomas replied:
If a user has enabled OpenTelemetryLoggerProvider (eg: https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/docs/logs/getting-started/Program.cs#L33), then this would be a never-ending-loop.

cijothomas replied:
@ejsmith Could you check this please?
(Sorry I haven't yet reviewed the recent change, but this part was an open issue even before, and not sure if this was addressed).

ejsmith replied:
@cijothomas I just tool a look at this and the `Examples.AspNetCore` project already uses the OTel logging provider and it works fine. It does not result in a never-ending-loop.

cijothomas commented on src/OpenTelemetry.Extensions.Hosting/Implementation/SelfDiagnosticsLoggingHostedService.cs:
```diff
@@ -0,0 +1,54 @@
+
+namespace OpenTelemetry.Extensions.Hosting.Implementation
+{
+    internal class SelfDiagnosticsLoggingHostedService : IHostedService
```
Adding background hosted services is not supported in some environments like Azure functions. (https://github.com/open-telemetry/opentelemetry-dotnet/pull/1672#discussion_r551650928)

The original reason why we used IHostedService was to have someone try to retrieve "TracerProvider" from DI, to trigger its creation. It *might* be possible for us to get rid of IHostedService usage in that scenario.

Adding more will make it difficult.

ejsmith replied:
I'd prefer to have this just be added as an instrumentation in the TracerProvider and have the same lifetime as that class. But unfortunately I can't access the IServiceProvider in order to get access to the ILoggerFactory from the TracerProviderBuilder. Which is why I was asking if you would be interested in a PR for that. https://github.com/open-telemetry/opentelemetry-dotnet/issues/894

ejsmith commented:
Changed to draft to wait for https://github.com/open-telemetry/opentelemetry-dotnet/pull/1889

ejsmith commented:
This is up to date with main now and I think it's ready to be considered for merging. It needs to live in it's own hosted service or it won't be able to log startup issues.

CodeBlanch commented on src/OpenTelemetry.Extensions.Hosting/Implementation/EventSourceEventFormatting.cs:
```diff
@@ -0,0 +1,93 @@
+
+namespace OpenTelemetry.Internal
+{
+    internal static class EventSourceEventFormatting
```
nit: `EventSourceEventFormatting` sounds more like a namespace, maybe `EventSourceEventFormatter` would be more standard? 🤷

ejsmith replied:
I agree with that. Changed.

CodeBlanch commented on src/OpenTelemetry.Extensions.Hosting/Implementation/EventSourceEventFormatting.cs:
```diff
@@ -0,0 +1,93 @@
+
+        public static string Format(EventWrittenEventArgs eventData)
+        {
+            var payloadArray = eventData.Payload?.ToArray() ?? Array.Empty<object>();
```
I feel like this array allocation + copy could be avoided. Payload is a ReadOnlyCollection<object> already I think? Could we do...

```csharp
private static readonly ReadOnlyCollection<object> emptyPayload = new ReadOnlyCollection<object>(Array.Empty<object>());

var payloadCollection = eventData.Payload ?? emptyPayload;
```

I think the for loops don't need to change, there should be count + indexer on ReadOnlyCollection.

ejsmith replied:
The items are being formatted and updated in `ProcessPayloadArray` so it can't be a read only collection.

CodeBlanch commented on src/OpenTelemetry.Extensions.Hosting/Implementation/EventSourceEventFormatting.cs:
```diff
@@ -0,0 +1,93 @@
+                }
+            }
+
+            var stringBuilder = Builder;
```
Will Format ever be called by more than one thread? Builder would get way corrupt in that case 😄

ejsmith replied:
Yeah, that could very well be an issue. I originally changed it to use a shared StringBuilder based on feedback from @reyang. I think what would be ideal is to make use of an ObjectPool, but we would need to bring in a package or do a source import to have access to an ObjectPool. It does seem like something that could be useful throughout this library.

ejsmith replied:
I just pushed a change that adds a simple `StringBuilderPool`.

ejsmith commented:
@CodeBlanch @cijothomas @reyang I added tests and I think I've addressed all the feedback. Can you please take a look at this and let me know what needs to be done in order to get this merged?

cijothomas commented on test/OpenTelemetry.Extensions.Hosting.Tests/OpenTelemetry.Extensions.Hosting.Tests.csproj:
```diff
@@ -9,11 +9,13 @@
   <ItemGroup>
     <PackageReference Include="Microsoft.NET.Test.Sdk" Version="$(MicrosoftNETTestSdkPkgVer)" />
     <PackageReference Include="Moq" Version="$(MoqPkgVer)" />
+    <PackageReference Include="Foundatio.Xunit" Version="$(FoundatioXUnitPkgVer)" />
```
Is there a reason why we want to add new packages for testing?

ejsmith replied:
Yes, it's an ILoggerFactory adaptor for Xunit that also allows gathering the log messages in memory and inspecting them. In this case I'm using it to verify that the expected log messages are being output.

cijothomas commented on src/OpenTelemetry.Extensions.Hosting/Implementation/SelfDiagnosticsLoggingHostedService.cs:
```diff
@@ -0,0 +1,54 @@
+            // The sole purpose of this HostedService is to
+            // start forwarding the self-diagnostics events
+            // to the logger factory
+            this.forwarder = new SelfDiagnosticsEventLogForwarder(System.Diagnostics.Tracing.EventLevel.LogAlways, this.serviceProvider.GetService<ILoggerFactory>());
```
`LogAlways` here means all the logs are going to collected by the forwarder.... Won't it mean checks like this (https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Internal/OpenTelemetrySdkEventSource.cs#L95), would always be true and causes perf issues?

ejsmith replied:
The messages get filtered out in the logger and can be controlled by logging configuration. The same logging system used in all of .NET Core production apps. I could make this a parameter going into the addselfdiagnostics method, but that would mean you would need to manually set that value and changing it in your logging configuration might not have any affect.

johnduhart replied:
The specification calls out having a dedicated option for controlling log level, so I think being able to adjust that independently of an application's logging level makes sense. https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/sdk-environment-variables.md#general-sdk-configuration

cijothomas replied:
Even if the ILogger logs are configured for ,say, only Error level, the current code subscribes to *all level* EventsSource logs (and drop the ones below Error.). This'd affect perf.

ejsmith replied:
@cijothomas I just pushed changes that set the EventLevels based on the logging configuration. It will also update them if the logging configuration changes at run time.

cijothomas replied:
Thanks. Will re-review this part.

cijothomas reviewed (changes requested):
While this is trying to solve a reasonable ask (forward logs to Asp.Net Core's Ilogger mechanism), there are open questions unresolved, hence requesting-changes until we resolve them.

ejsmith commented:
To me this isn't a reasonable ask. It's mandatory and what I would expect. It is currently very hard to figure out what is going on when I've misconfigured something and my APM data is just disappearing and not making it to the destination. Also, when I'm first starting out it's very valuable to see how things work.

ejsmith commented:
In production if I have the log level set to error then I can see when bad things are happening inside of Otel instead of them just being eaten and I never know about them.

CodeBlanch commented on src/OpenTelemetry.Extensions.Hosting/Implementation/SelfDiagnosticsLoggingHostedService.cs:
```diff
@@ -0,0 +1,55 @@
+            // The sole purpose of this HostedService is to
+            // start forwarding the self-diagnostics events
+            // to the logger factory
+            this.forwarder = new SelfDiagnosticsEventLogForwarder(this.serviceProvider.GetService<ILoggerFactory>(), this.serviceProvider.GetService<IOptionsMonitor<LoggerFilterOptions>>(), System.Diagnostics.Tracing.EventLevel.Warning);
```
@ejsmith Just curious here why you are calling `this.serviceProvider.GetService` couldn't you inject those services via ctor?

ejsmith replied:
Yeah, you are right. They could probably be injected.

ejsmith replied:
Updated

CodeBlanch commented on src/OpenTelemetry.Extensions.Hosting/OpenTelemetryServicesExtensions.cs:
```diff
@@ -55,6 +56,24 @@ public static IServiceCollection AddOpenTelemetryTracing(this IServiceCollection
+        /// </remarks>
+        /// <param name="services">The <see cref="IServiceCollection" /> to add services to.</param>
+        /// <returns>The <see cref="IServiceCollection"/> so that additional calls can be chained.</returns>
+        public static IServiceCollection AddOpenTelemetrySelfDiagnosticsLogging(this IServiceCollection services)
```
@ejsmith Just curious why you didn't use the new DI extensions here. I'm guessing because that would work for TracerProviderBuilder but this is a kind of universal thing for tracing, metrics, & logging?

ejsmith replied:
@CodeBlanch that was the idea originally and I was waiting for it, but the problem is that if I use those then I don't capture otel startup issues in the logging because it's not setup until after startup happens.

CodeBlanch replied:
Ah order issue. I think in `TelemetryHostedService` the `TracerProvider` is created in `StartAsync` so if in `SelfDiagnosticsLoggingHostedService` you got everything up and running in the ctor instead of `StartAsync` it might work and capture everything. But that's more academic you don't have to change anything unless you want to.

ejsmith replied:
@CodeBlanch good idea. I changed it around and I like it better being an extension on the TraceProviderBuilder.

ejsmith commented:
@cijothomas @CodeBlanch can I please get an update on the status of this PR?

ejsmith commented:
> This PR was marked stale due to lack of activity. It will be closed in 7 days.

Sigh.
